src/Backends/PrefersColorSchemeSettings.vala

```diff
@@ -0,0 +1,151 @@
+* General Public License for more details.
+*
+* You should have received a copy of the GNU General Public
+* License along with this program; if not, write to the
+* Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+* Boston, MA 02110-1301 USA
+*
+* Authored by: Marius Meisenzahl <mariusmeisenzahl@gmail.com>
+*/
+
+public class SettingsDaemon.Backends.PrefersColorSchemeSettings : GLib.Object {
+    public unowned PantheonShell.Pantheon.AccountsService accounts_service { get; construct; }
+
+    private GLib.Settings color_settings;
+    private double sunrise = -1.0;
+    private double sunset = -1.0;
+
+    private uint time_id = 0;
+
+    public PrefersColorSchemeSettings (PantheonShell.Pantheon.AccountsService accounts_service) {
+        Object (accounts_service: accounts_service);
+    }
+
+    construct {
+        color_settings = new GLib.Settings ("io.elementary.settings-daemon.prefers-color-scheme");
+
+        color_settings.changed["prefer-dark-schedule"].connect (update);
+
+        var variant = color_settings.get_value ("last-coordinates");
+        on_location_updated (variant.get_child_value (0).get_double (), variant.get_child_value (1).get_double ());
+
+        update ();
+    }
+
+    private void update () {
+        var schedule = color_settings.get_string ("prefer-dark-schedule");
+
+        if (schedule == "sunset-to-sunrise") {
+            get_location.begin ();
+
+            start_timer ();
+        } else if (schedule == "manual") {
+            start_timer ();
+        } else {
+            stop_timer ();
+        }
+    }
+
+    private void start_timer () {
+        if (time_id == 0) {
+            var time = new TimeoutSource (1000);
+            time.set_callback (time_callback);
+            time_id = time.attach (null);
+        }
+    }
+
+    private void stop_timer () {
+        if (time_id != 0) {
+            Source.remove (time_id);
+            time_id = 0;
+        }
+    }
+
+    private async void get_location () {
+        try {
+            var simple = yield new GClue.Simple (Build.PROJECT_NAME, GClue.AccuracyLevel.CITY, null);
+
+            simple.notify["location"].connect (() => {
+                on_location_updated (simple.location.latitude, simple.location.longitude);
+            });
+
+            on_location_updated (simple.location.latitude, simple.location.longitude);
+        } catch (Error e) {
+            warning ("Failed to connect to GeoClue2 service: %s", e.message);
+            return;
+        }
+    }
+
+    private bool time_callback () {
+        var schedule = color_settings.get_string ("prefer-dark-schedule");
+
+        var now = new DateTime.now_local ();
+        double from, to;
+        if (schedule == "sunset-to-sunrise") {
+            if (sunrise >= 0 && sunset >= 0) {
+                from = sunset;
+                to = sunrise;
+            } else {
+                from = 20.0;
```

Comment:
Could you add a comment here to note that these are fallback times (6AM and 8PM) for when an invalid result was returned from the calculation (i.e. probably wasn't able to get a location).